modules/infra/control/control_output.c

```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2024 Christophe Fontaine
+
+#include <gr_control.h>
+#include <gr_control_output.h>
+#include <gr_graph.h>
+#include <gr_log.h>
+#include <gr_macro.h>
+#include <gr_mbuf.h>
+
+#include <event2/event.h>
+#include <rte_ether.h>
+#include <rte_graph_worker.h>
+
+#include <pthread.h>
+#include <stdatomic.h>
+#include <stdlib.h>
+
+extern struct rte_ring *control_output_ring;
```

Comment:
This ring should be a static variable in this module. And preferably with a shorter name. You should expose a public function to enqueue packets from dataplane without exposing the ring variable.